File: fix_marquees.py
import os
import re
import glob

services_dir = 'services'
html_files = glob.glob(os.path.join(services_dir, '*.html'))

def replace_marquee(match):
    content = match.group(0)
    # find row-right and row-left contents
    def duplicate_content(row_content):
        # find the inner spans
        spans = re.findall(r'<span class="tech-chip.*?</span>', row_content)
        # assuming the original set has 4 unique chips, and it was duplicated once to 8 chips.
        # let's just grab the first 4 chips
        if len(spans) >= 4:
            base_spans = spans[:(len(spans)//2)]
            # duplicate it so we have 4 sets (e.g. 16 chips if base is 4)
            return ' '.join(base_spans * 4)
        return row_content
    
    # regex to find the inner content of div.tech-ticker-row
    new_content = re.sub(
        r'(<div class="tech-ticker-row row-right">)(.*?)(</div>)',
        lambda m: m.group(1) + duplicate_content(m.group(2)) + m.group(3),
        content,
        flags=re.DOTALL
    )
    new_content = re.sub(
        r'(<div class="tech-ticker-row row-left">)(.*?)(</div>)',
        lambda m: m.group(1) + duplicate_content(m.group(2)) + m.group(3),
        new_content,
        flags=re.DOTALL
    )
    return new_content

for file in html_files:
    with open(file, 'r', encoding='utf-8') as f:
        html = f.read()
    
    # replace inside tech-marquee-wrapper
    new_html = re.sub(
        r'<div class="tech-marquee-wrapper">.*?</div>\s*</div>\s*</section>',
        lambda m: replace_marquee(m),
        html,
        flags=re.DOTALL
    )
    
    with open(file, 'w', encoding='utf-8') as f:
        f.write(new_html)
    print(f"Updated {file}")

File: fix_marquees.js
const fs = require('fs');
const path = require('path');

const servicesDir = 'services';
const files = fs.readdirSync(servicesDir).filter(f => f.endsWith('.html'));

function replaceMarquee(content) {
    const rowRightRegex = /(<div class="tech-ticker-row row-right">)(.*?)(<\/div>)/s;
    const rowLeftRegex = /(<div class="tech-ticker-row row-left">)(.*?)(<\/div>)/s;
    
    function duplicateContent(rowStr) {
        let spans = [];
        const regex = /<span class="tech-chip.*?<\/span>/gs;
        let match;
        while ((match = regex.exec(rowStr)) !== null) {
            spans.push(match[0]);
        }
        if (spans.length >= 4) {
            let base = spans.slice(0, spans.length / 2); // original duplicated once
            let combined = '';
            for(let i=0; i<4; i++){
                combined += '\\n        ' + base.join('\\n        ') + '\\n';
            }
            return combined;
        }
        return rowStr;
    }

    let newContent = content;
    
    newContent = newContent.replace(rowRightRegex, (match, p1, p2, p3) => {
        return p1 + duplicateContent(p2) + p3;
    });

    newContent = newContent.replace(rowLeftRegex, (match, p1, p2, p3) => {
        return p1 + duplicateContent(p2) + p3;
    });

    return newContent;
}

for (const file of files) {
    const fullPath = path.join(servicesDir, file);
    let html = fs.readFileSync(fullPath, 'utf8');
    
    const sectionRegex = /(<div class="tech-marquee-wrapper">.*?<\/div>\s*<\/div>\s*<\/section>)/s;
    html = html.replace(sectionRegex, (match) => {
        return replaceMarquee(match);
    });
    
    fs.writeFileSync(fullPath, html, 'utf8');
    console.log(`Updated ${file}`);
}
